src/cargo/ops/cargo_add/mod.rs

```diff
@@ -943,56 +943,57 @@ fn print_dep_table_msg(shell: &mut Shell, dep: &DependencyUI) -> CargoResult<()>
-
-        const MAX_FEATURE_PRINTS: usize = 50;
-
-        let mut activated_printed = 0;
-        let total_activated = activated.len();
-        for feat in activated {
-            if activated_printed >= MAX_FEATURE_PRINTS {
-                let remaining = total_activated - activated_printed;
-                shell.write_stderr(
-                    format_args!("{prefix}... {remaining} more activated features\n"),
-                    &style::NOP,
-                )?;
-                break;
-            }
+    if activated.is_empty() && deactivated.is_empty() {
+        return Ok(());
+    }
 
-            shell.write_stderr(&prefix, &style::NOP)?;
-            shell.write_stderr('+', &style::GOOD)?;
-            shell.write_stderr(format_args!(" {}\n", feat), &style::NOP)?;
-            activated_printed += 1;
-        }
+    let prefix = format!("{:>13}", " ");
+    let suffix = format_features_version_suffix(&dep);
 
-        let mut deactivated_printed = 0;
-        let total_deactivated = deactivated.len();
-        for feat in deactivated {
-            if activated_printed + deactivated_printed >= MAX_FEATURE_PRINTS {
-                let remaining = total_deactivated - deactivated_printed;
-                shell.write_stderr(
-                    format_args!("{prefix}... {remaining} more deactivated features\n"),
-                    &style::NOP,
-                )?;
-                break;
-            }
+    shell.write_stderr(format_args!("{prefix}Features{suffix}:\n"), &style::NOP)?;
 
-            shell.write_stderr(&prefix, &style::NOP)?;
-            shell.write_stderr('-', &style::ERROR)?;
-            shell.write_stderr(format_args!(" {}\n", feat), &style::NOP)?;
-            deactivated_printed += 1;
-        }
+    const MAX_FEATURE_PRINTS: usize = 30;
+    let total_activated = activated.len();
+    let total_deactivated = deactivated.len();
+
+    if total_activated > MAX_FEATURE_PRINTS {
+        shell.write_stderr(
+            format_args!("{prefix}{total_activated} activated features\n"),
+            &style::NOP,
+        )?;
+        shell.write_stderr(
+            format_args!("{prefix}{total_deactivated} deactivated features\n"),
+            &style::NOP,
+        )?;
+        return Ok(());
+    }
+
+    for feat in activated {
+        shell.write_stderr(&prefix, &style::NOP)?;
+        shell.write_stderr('+', &style::GOOD)?;
+        shell.write_stderr(format_args!(" {feat}\n"), &style::NOP)?;
+    }
+
+    if total_activated + total_deactivated > MAX_FEATURE_PRINTS {
+        shell.write_stderr(
+            format_args!("{prefix}{total_deactivated} deactivated features\n"),
+            &style::NOP,
+        )?;
+        return Ok(());
+    }
+
+    for feat in deactivated {
+        shell.write_stderr(&prefix, &style::NOP)?;
+        shell.write_stderr('-', &style::ERROR)?;
+        shell.write_stderr(format_args!(" {feat}\n"), &style::NOP)?;
     }
```

Comment:
For me, early returns are for guard statements / boilerplate.  When it involves business logic, `else`s should be used.  This is on top of the problem of bugs if/when new features are added as pointed out in m,y other comment.

This can be as simple as:
```rust
if activated.len() <MAX_FEATURE_PRINTS {
    ... for loop
} else {
    .. summary
}
if total_activated + total_deactivated < MAX_FEATURE_PRINTS {
    ... for loop
} else {
    ... summary
}
```

Reply:
Agreed for this case